keps/sig-auth/3325-self-subject-attributes-review-api/README.md

```diff
@@ -0,0 +1,368 @@
+The motivation for this KEP is to reduce obscurity and help users with debugging the authentication stack.
+
+### Goals
+
+- Add the API endpoint to get user attributes
+- Add a corresponding kubectl command - `kubectl auth who-am-i`
+
+### Non-Goals
+
+## Proposal
+
+Add a new API endpoint to the `authentication` group - `SelfSubjectAttributesReview`.
+The user will hit the endpoint after authentication happens, so all attributes will be available to return.
+
+## Design Details
+
+This design is inspired by the `*AccessReview` and `TokenReview` APIs.
+The endpoint has no input parameters or a `spec` field because only the authentication result is required.
+
+### Request
+
+The structure for building a request:
+```go
+type SelfSubjectAttributesReview struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status is filled in by the server with the user attributes.
+	Status SelfSubjectAttributesReview `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+```
+```go
+type SelfSubjectAttributesReviewStatus struct {
+	// User attributes of the current user.
+	// +optional
+	UserInfo UserInfo `json:"userInfo,omitempty" protobuf:"bytes,1,opt,name=userInfo"`
+}
+
+type UserInfo struct {
+	Name   string              `json:"name" protobuf:"bytes,1,opt,name=name"`
+	UID    string              `json:"uid" protobuf:"bytes,2,opt,name=uid"`
+	Groups []string            `json:"groups" protobuf:"bytes,3,opt,name=groups"`
+	Extra  map[string][]string `json:"extra" protobuf:"bytes,4,opt,name=extra"`
+}
+```
+
+On receiving a request, the Kubernetes API server fills the status with the user attributes and returns it to the user.
+
+Request example (the body would be a `SelfSubjectAttributesReview` object):
+```
+POST /apis/authentication.k8s.io/v1alpha1/selfsubjectattributesreview
+```
+```json
+{
+  "apiVersion": "authentication.k8s.io/v1alpha1",
+  "kind": "SelfSubjectAttributesReview"
+}
+```
+Response example:
+
+```json
+{
+  "apiVersion": "authentication.k8s.io/v1alpha1",
+  "kind": "SelfSubjectAttributesReview",
+  "status": {
+    "name": "jane.doe",
+    "uid": "b6c7cfd4-f166-11ec-8ea0-0242ac120002",
+    "groups": ["viewers", "editors"],
+    "extra": {
+      "provider_id": "token.company.dev"
+    }
+  }
+}
+```
+
+User attributes are known at the moment of accessing the rest API endpoint and can be extracted from the request context.
+
+NOTE: There are no audiences in requests and responses since the SelfSubjectAttributesReview API is implied to be simple.
+Unlike the TokenReview API works, kube-apiserver will not do additional internal requests.
+Instead, a user will see the exact result of the authentication, which will be extracted from the request context.
```

Comment:
The reason I believe audience should not be included is because the only way to access this API is via a credential that is valid against the API server itself, meaning that the audience must always be that of the API server (or the credential is audience agnostic).  Either way, the specific value of the audience is irrelevant.  Learning this value is not useful.